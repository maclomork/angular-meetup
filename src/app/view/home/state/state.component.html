<div class="container">

  <div class="row">
    <div class="col-12">
      <h1>Buttons Clicked State</h1>
    </div>
    <div class="col-6">
      <p id="left-btn-clicks">Left Button clicked: {{leftButtonClicked$ | async}}</p>
    </div>
    <div class="col-6">
      <p id="right-btn-clicks">Right Button clicked: {{rightButtonClicked$ | async}}</p>
    </div>
  </div>

  <div class="row">
    <div class="col-12">
      <h1>Input Form State</h1>
    </div>
    <div class="col-12">
      <p id="form-ctrl-valid">FormControl valid: {{formControlValid$ | async}}</p>
    </div>
    <div class="col-12">
      <p id="form-ctrl-value">FormControl value: {{formControlValue$ | async}}</p>
    </div>
  </div>

  <div class="row">
    <div class="col-12">
      <h1>Checkbox State</h1>
    </div>
    <div class="col-12">
      <p id="checkbox-value">Checkbox Value: {{checkBoxValue$ | async}}</p>
    </div>
  </div>

</div>
