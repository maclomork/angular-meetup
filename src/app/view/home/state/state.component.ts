import {Component, OnInit} from '@angular/core';
import {StateService} from '../../../shared/state.service';

@Component({
  selector: 'app-state',
  templateUrl: './state.component.html',
  styleUrls: ['./state.component.scss']
})
export class StateComponent implements OnInit {

  leftButtonClicked: number;
  rightButtonClicked: number;

  formControlValid: boolean;
  formControlValue: string;

  checkBoxValue: boolean;

  constructor(private stateService: StateService) {
    this.leftButtonClicked = 0;
    this.rightButtonClicked = 0;
    this.formControlValid = false;
    this.formControlValue = '';
    this.checkBoxValue = false;
  }

  ngOnInit() {
    this.stateService.leftButtonClicksSubject.subscribe(leftBtnClicked => this.leftButtonClicked = leftBtnClicked);
    this.stateService.rightButtonClicksSubject.subscribe(rightBtnClicked => this.rightButtonClicked = rightBtnClicked);

    this.stateService.formControlValidSubject.subscribe(formCtrlValid => this.formControlValid = formCtrlValid);
    this.stateService.formControlValueSubject.subscribe(formCtrlValue => this.formControlValue = formCtrlValue);

    this.stateService.checkBoxValue.subscribe(checkBoxValue => this.checkBoxValue = checkBoxValue);
  }

}
