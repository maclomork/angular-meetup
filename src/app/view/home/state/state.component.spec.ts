import {async, ComponentFixture, TestBed} from '@angular/core/testing';

import {StateComponent} from './state.component';
import {StateService} from '../../../shared/state.service';
import {Subject} from 'rxjs';

class StateServiceMock {
  leftButtonClicksSubject: Subject<number> = new Subject<number>();
  rightButtonClicksSubject: Subject<number> = new Subject<number>();

  formControlValidSubject: Subject<boolean> = new Subject<boolean>();
  formControlValueSubject: Subject<string> = new Subject<string>();

  checkBoxValueSubject: Subject<boolean> = new Subject<boolean>();
}

describe('StateComponent', () => {
  let component: StateComponent;
  let fixture: ComponentFixture<StateComponent>;
  let stateServiceMock: StateServiceMock;

  beforeEach(async(() => {
    stateServiceMock = new StateServiceMock();

    TestBed.configureTestingModule({
      declarations: [StateComponent],
      providers: [{
        provide: StateService,
        useValue: stateServiceMock
      }]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(StateComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('StateComponent can be created', () => {
    expect(component).toBeTruthy();
  });

  test('StateComponent displays the amount of button clicks', () => {
    const leftBtnClicksElement = fixture.debugElement.nativeElement.querySelector('#left-btn-clicks');
    const rightBtnClicksElement = fixture.debugElement.nativeElement.querySelector('#right-btn-clicks');

    expect(component.leftButtonClicked).toBe(0);
    expect(leftBtnClicksElement.textContent).toEqual('Left Button clicked: 0');
    expect(component.rightButtonClicked).toBe(0);
    expect(rightBtnClicksElement.textContent).toEqual('Right Button clicked: 0');

    stateServiceMock.leftButtonClicksSubject.next(5);
    stateServiceMock.rightButtonClicksSubject.next(15);

    fixture.detectChanges();

    expect(component.leftButtonClicked).toBe(5);
    expect(leftBtnClicksElement.textContent).toEqual('Left Button clicked: 5');
    expect(component.rightButtonClicked).toBe(15);
    expect(rightBtnClicksElement.textContent).toEqual('Right Button clicked: 15');
  });

  test('StateComponent displays the information about input form control', () => {
    const formCtrlValidElement = fixture.debugElement.nativeElement.querySelector('#form-ctrl-valid');
    const formCtrlValueElement = fixture.debugElement.nativeElement.querySelector('#form-ctrl-value');

    expect(component.formControlValid).toBe(false);
    expect(formCtrlValidElement.textContent).toEqual('FormControl valid: false');
    expect(component.formControlValue).toBe('');
    expect(formCtrlValueElement.textContent).toEqual('FormControl value: ');

    stateServiceMock.formControlValidSubject.next(true);
    stateServiceMock.formControlValueSubject.next('some text');

    fixture.detectChanges();

    expect(component.formControlValid).toBe(true);
    expect(formCtrlValidElement.textContent).toEqual('FormControl valid: true');
    expect(component.formControlValue).toBe('some text');
    expect(formCtrlValueElement.textContent).toEqual('FormControl value: some text');
  });

  test('StateComponent displays the checkbox form control value', () => {
    const checkBoxValueElement = fixture.debugElement.nativeElement.querySelector('#checkbox-value');

    expect(component.checkBoxValue).toBe(false);
    expect(checkBoxValueElement.textContent).toEqual('Checkbox Value: false');

    stateServiceMock.checkBoxValueSubject.next(true);

    fixture.detectChanges();

    expect(component.checkBoxValue).toBe(true);
    expect(checkBoxValueElement.textContent).toEqual('Checkbox Value: true');
  });

  test('Iterating the subjects results in correct values displayed', () => {
    stateServiceMock.leftButtonClicksSubject.next(4);
    stateServiceMock.rightButtonClicksSubject.next(12);
    stateServiceMock.formControlValueSubject.next('some text');
    stateServiceMock.formControlValidSubject.next(true);
    stateServiceMock.checkBoxValueSubject.next(true);

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
