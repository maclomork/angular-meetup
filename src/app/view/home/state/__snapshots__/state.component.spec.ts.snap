// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateComponent Iterating the subjects results in correct values displayed 1`] = `
<app-state
  checkBoxValue={[Function Boolean]}
  formControlValid={[Function Boolean]}
  formControlValue={[Function String]}
  leftButtonClicked={[Function Number]}
  rightButtonClicked={[Function Number]}
  stateService={[Function StateServiceMock]}
>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <h1>
          Buttons Clicked State
        </h1>
      </div>
      <div
        class="col-6"
      >
        <p
          id="left-btn-clicks"
        >
          Left Button clicked: 4
        </p>
      </div>
      <div
        class="col-6"
      >
        <p
          id="right-btn-clicks"
        >
          Right Button clicked: 12
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <h1>
          Input Form State
        </h1>
      </div>
      <div
        class="col-12"
      >
        <p
          id="form-ctrl-valid"
        >
          FormControl valid: true
        </p>
      </div>
      <div
        class="col-12"
      >
        <p
          id="form-ctrl-value"
        >
          FormControl value: some text
        </p>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-12"
      >
        <h1>
          Checkbox State
        </h1>
      </div>
      <div
        class="col-12"
      >
        <p
          id="checkbox-value"
        >
          Checkbox Value: true
        </p>
      </div>
    </div>
  </div>
</app-state>
`;
